File: src/test/kotlin/com/djulia/transactions/ResultCapturingMessageConverter.java
package com.djulia.transactions;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.http.HttpInputMessage;
import org.springframework.http.HttpOutputMessage;
import org.springframework.http.MediaType;
import org.springframework.http.converter.HttpMessageConverter;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.http.converter.HttpMessageNotWritableException;

import java.io.IOException;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

class ResultCapturingMessageConverter implements HttpMessageConverter<Object> {
    private Optional<?> result;

    @Override
    public boolean canRead(Class<?> clazz, MediaType mediaType) {
        return true;
    }

    @Override
    public boolean canWrite(Class<?> clazz, MediaType mediaType) {
        return true;
    }

    @Override
    public List<MediaType> getSupportedMediaTypes() {
        return Collections.singletonList(MediaType.ALL);
    }

    @Override
    public Object read(Class<?> clazz, HttpInputMessage inputMessage) throws IOException, HttpMessageNotReadableException {
       return new ObjectMapper().readValue(inputMessage.getBody(), clazz);
    }

    @Override
    public void write(Object o, MediaType contentType, HttpOutputMessage outputMessage) throws IOException, HttpMessageNotWritableException {
        result = Optional.of(o);
//        new ObjectMapper().writeValue(outputMessage.getBody(), o);
    }

    public Object getResult() {
        return result.orElseThrow(
                () -> new RuntimeException("Oops." +
                        "Are you sure you didn't forget to invoke the controller method in MockMvc?")
        );
    }
}

File: transactions/src/test/com/djulia/banking/transactions/ResultCapturingMessageConverter.kt
package com.djulia.banking.transactions

import org.springframework.http.HttpInputMessage
import org.springframework.http.HttpOutputMessage
import org.springframework.http.MediaType
import org.springframework.http.converter.HttpMessageConverter
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.http.converter.HttpMessageNotWritableException
import java.io.IOException

internal class ResultCapturingMessageConverter : HttpMessageConverter<Any> {
    private var result: Any? = null

    override fun canRead(clazz: Class<*>, mediaType: MediaType?): Boolean {
        return true
    }

    override fun canWrite(clazz: Class<*>, mediaType: MediaType?): Boolean {
        return true
    }

    override fun getSupportedMediaTypes(): List<MediaType> {
        return listOf(MediaType.ALL)
    }

    @Throws(IOException::class, HttpMessageNotReadableException::class)
    override fun read(clazz: Class<*>, inputMessage: HttpInputMessage): Any {
        return makeDefaultObjectMapper().readValue(inputMessage.body, clazz)
    }

    @Throws(IOException::class, HttpMessageNotWritableException::class)
    override fun write(o: Any, contentType: MediaType, outputMessage: HttpOutputMessage) {
        result = o
    }

    fun getResult(): Any {
        return result ?: throw RuntimeException("Oops." + "Are you sure you didn't forget to invoke the controller method in MockMvc?")
    }
}
